jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/MetaInfConfiguration.java

```diff
@@ -589,58 +592,46 @@ public void postConfigure(WebAppContext context) throws Exception
         context.setAttribute(METAINF_TLDS, null);
     }
 
-    /**
-     * Find all .tld files in all subdirs of the given dir.
-     *
-     * @param dir the directory to scan
-     * @return the list of tlds found
-     * @throws IOException if unable to scan the directory
-     */
-    private Collection<URL> getTlds(Path dir) throws IOException
-    {
-        if (dir == null || !Files.isDirectory(dir))
-            return Collections.emptySet();
-
-        Set<URL> tlds = new HashSet<>();
-
-        try (Stream<Path> entries = Files.walk(dir)
-            .filter(Files::isRegularFile)
-            .filter(FileID::isTld))
-        {
-            Iterator<Path> iter = entries.iterator();
-            while (iter.hasNext())
-            {
-                Path entry = iter.next();
-                tlds.add(entry.toUri().toURL());
-            }
-        }
-        return tlds;
-    }
-
     /**
      * Find all .tld files in the given jar.
      *
-     * @param uri the uri to jar file
+     * @param resource the jar file
      * @return the collection of tlds as url references
      * @throws IOException if unable to scan the jar file
      */
-    private Collection<URL> getTlds(WebAppContext context, URI uri) throws IOException
+    private Collection<URL> getTlds(WebAppContext context, Resource resource) throws IOException
     {
         HashSet<URL> tlds = new HashSet<>();
-        Resource r = ResourceFactory.of(context).newResource(URIUtil.uriJarPrefix(uri, "!/"));
-        try (Stream<Path> stream = Files.walk(r.getPath()))
+        Resource resourceDir;
+
+        try (ResourceFactory.Closeable closeableResourceFactory = ResourceFactory.closeable())
         {
-            Iterator<Path> it = stream
-                .filter(Files::isRegularFile)
-                .filter(FileID::isTld)
-                .iterator();
-            while (it.hasNext())
+            if (!resource.isDirectory())
             {
-                Path entry = it.next();
-                tlds.add(entry.toUri().toURL());
+                //turn it into a jar:file
+                resourceDir = closeableResourceFactory.newJarFileResource(resource.getURI());
+            }
+            else
+                resourceDir = resource;
+
+            Resource metaInf = resourceDir.resolve("META-INF");
+            if (!metaInf.isDirectory())
+                return tlds; //no tlds
+
+            try (Stream<Path> stream = Files.walk(metaInf.getPath()))
```

Comment:
Use the Resource API, not Path + Walk. (like the other PR does).

In fact, I just checked, both this PR and my other one does this entire step (scan for TLDs) incorrectly per spec.

For webapps, its a recursive scan for anything `*.tld` from root.
* ignore if found in `/WEB-INF/classes/`
* ignore if found in `/WEB-INF/lib/`
* ignore if ends in `/` (ie: it's a directory called `foo.tld/`)
* ignore if found in `/WEB-INF/tags/` (with exception for `implicit.tld`)
* all other hits are valid.

Reply:
This can't be correct, as if you look at all previous versions of jetty that are looking for `tlds` we're using the `JarFile` class to enumerate all entries in the jar, then picking those that start with `META_INF` and end with `.tld`.

Reply:
I opened Issue #10899 for this (with links to relevant parts of the spec)